internal/rootcoord/meta_table.go

```diff
@@ -973,6 +1037,138 @@ func (mt *MetaTable) GetInitBuildIDs(collName, indexName string) ([]UniqueID, er
 	return initBuildIDs, nil
 }
 
+func (mt *MetaTable) AlignSegmentsMeta(collID, partID UniqueID, segIDs map[UniqueID]struct{}) {
+	mt.ddLock.Lock()
+
+	allIndexID := make([]UniqueID, 0)
+	if collMeta, ok := mt.collID2Meta[collID]; ok {
+		for _, fieldIndex := range collMeta.FieldIndexes {
+			allIndexID = append(allIndexID, fieldIndex.IndexID)
+		}
+	}
+
+	recycledSegIDs := make([]UniqueID, 0)
+	if segMap, ok := mt.partID2SegID[partID]; ok {
+		for segID := range segMap {
+			if _, ok := segIDs[segID]; !ok {
+				recycledSegIDs = append(recycledSegIDs, segID)
+			}
+		}
+	}
+
+	mt.ddLock.Unlock()
```

Comment:
the lock handling is weird.
Usually we can RemoveSegments and RemoveSegmentsInternal where RemoveSegmentsInternal don't get lock